src/FHIR/SMART/SmartLaunchController.php

```diff
@@ -0,0 +1,155 @@
+use OpenEMR\Common\Auth\OpenIDConnect\Repositories\ClientRepository;
+use OpenEMR\Common\Uuid\UuidRegistry;
+use OpenEMR\Events\PatientDemographics\RenderEvent;
+use OpenEMR\RestControllers\AuthorizationController;
+use OpenEMR\Services\PatientService;
+use Symfony\Component\EventDispatcher\EventDispatcher;
+
+// not sure I really like this here... it seems like some of this
+// should be encapsulated in a class that autoloading can reach.
+require_once(__DIR__ . '/../../../_rest_config.php');
+
+/**
+ * Class SmartLaunchController handles the display and launching of SMART apps from the user interface.
+ * @package OpenEMR\FHIR\SMART
+ */
+class SmartLaunchController
+{
+    /**
+     * @var EventDispatcher
+     */
+    private $dispatcher;
+
+    public function __construct(EventDispatcher $dispatcher = null)
+    {
+        $this->dispatcher = $dispatcher;
+    }
+
+    public function registerContextEvents()
+    {
+        $this->dispatcher->addListener(RenderEvent::EVENT_SECTION_LIST_RENDER_AFTER, [$this, 'renderPatientSmartLaunchSection']);
+    }
+
+    public function renderPatientSmartLaunchSection(RenderEvent $event)
+    {
+        $smartClients = $this->getSMARTClients();
+        // TODO: adunsulag we would filter the clients based on their smart capability & scopes they could send...
+        $pid = $event->getPid();
+        $patientService = new PatientService();
+        // going to work with string uuids
+        $puuid = UuidRegistry::uuidToString($patientService->getUuid($pid));
+        ?>
+        <section>
+            <?php
+            // Billing expand collapse widget
+            $widgetTitle = xl("SMART Enabled Apps");
+            $widgetLabel = "smart";
+            $widgetButtonLabel = xl("Edit");
+            $widgetButtonLink = ""; // "return newEvt();";
+            $widgetButtonClass = "";
+            $linkMethod = "javascript";
+            $bodyClass = "notab";
+            $widgetAuth = false;
+            $fixedWidth = false;
+            $forceExpandAlways = false;
+            $launchCode = $this->getLaunchCodeContext($puuid);
+            // TODO: adunsulag is there an redirect_uri that we can specify for the launch path?? The spec feels vague
+            // here...  all the SMART apps we've seen appear to follow a 'launch.html' nomenclature but that doesn't
+            // appear to be required in the spec.
+
+            $gbl = \RestConfig::GetInstance();
+            // TODO: adunsulag surely we can centralize where this fhir API url is set?
+            // TODO: adunsulag what is wrong with these URL's?  I'm having to hard code the issuer as I can't
+            // seem to get these URLs right.  for some reason the $SITE is set to interface, we don't get 'apis' in there
+            // ROOT_URL appears to be empty.. just strange
+            // $issuer = $GLOBALS['site_addr_oath'] . $gbl::$SITE . $gbl::$ROOT_URL . "/fhir";
+            $issuer = $GLOBALS['site_addr_oath'] . "/apis/default/fhir";
```

Comment:
is this supposed to be same issuer as oauth2 stuff?

Reply:
It's supposed to be the base FHIR url.  @sjpadgett said the best place to grab that was from site_addr_oath.

Reply:
This brings up another issue with the iss attribute issued by the OpenIDConnect server idTokenResponse. 
```
$builder = (new Builder())
            ->permittedFor($accessToken->getClient()->getIdentifier())
            ->issuedBy('https://' . $_SERVER['HTTP_HOST'])
            ->issuedAt(time())
            ->expiresAt($accessToken->getExpiryDateTime()->getTimestamp())
            ->relatedTo($userEntity->getIdentifier());

        return $builder;
``` 
Ideally all issuers should match with the way we have now where issuer is just the protocol/domain!
So for now `$issuer = $GLOBALS['site_addr_oath']`

Reply:
@sjpadgett , good to clarify this. now the issuer is also appending webroot:
```
$GLOBALS['site_addr_oath'] . $GLOBALS['webroot']
```
Should this then be changed to just:
```
$GLOBALS['site_addr_oath']
``

Reply:
yep. it hasn't caught me because I always alias into openemr. At least for OIDC spec that issuer in discovery must match any iss attribute in tokens. For our iss in id_token it comes from $_SERVER['HTTP_HOST'] so if webroot isn't passed in the super then we don't include in discovery.
Still we need to figure out how we can change where token builder gets its issuer. Then we can give out more meaningful issuers..

Reply:
So... I'm not following this completely, do I need to change this code then?

Reply:
In this case i'm not sure how issuer(iss) is used. If used to build the launch url then make it a full uri for the endpoint. If it becomes a claim then $issuer = $GLOBALS['site_addr_oath']
Does the app launch from the iss?
Here is the similar path from capability statement
'$GLOBALS['site_addr_oath'] . $gbl::$ROOT_URL . "/" . $gbl::$SITE . "/fhir"'

Reply:
@sjpadgett issuer here is the base fhir URL which the SMART app will use to download the capability statement AND/OR the smart well-known configuration.  So I guess you could say yes, it does launch from the iss.

I tried using that capability statement piece originally but $gbl::$ROOT_URL was empty.

Reply:
try to use $GLOBALS['web_root'] and $GLOBALS['site_id']

Reply:
somebody pulled $GLOBALS['site_id'] from restconfig so only option is $_SESSION['site_id'] if $gbl not in scope

Reply:
Yea, I just was going to post that I used the following to get this to work:
$issuer = $GLOBALS['site_addr_oath'] . $GLOBALS['web_root'] . '/apis/' . $_SESSION['site_id'] . "/fhir";

Reply:
okay. my bet is that 99% of clients out there won't bother with validating tokens. still I worry and in this case, iss is serving a diff purpose.
Still I need to fix oidc discovery